core/src/main/java/io/parsingdata/metal/expression/comparison/ComparisonExpression.java

```diff
@@ -37,11 +38,18 @@ public ComparisonExpression(final ValueExpression current, final ValueExpression
 
     @Override
     public boolean eval(final Environment env, final Encoding enc) {
-        final OptionalValue ocv = _current == null ? OptionalValue.of(env.order.current()) : _current.eval(env, enc);
-        if (!ocv.isPresent()) { return false; }
-        final OptionalValue opv = _predicate.eval(env, enc);
-        if (!opv.isPresent()) { return false; }
-        return compare(ocv.get(), opv.get());
+        final OptionalValueList ocl = _current == null ? OptionalValueList.create(OptionalValue.of(env.order.current())) : _current.eval(env, enc);
+        if (ocl.isEmpty()) { return false; }
+        final OptionalValueList opl = _predicate.eval(env, enc);
+        if (ocl.size != opl.size) { return false; }
+        return compare(ocl, opl);
+    }
+
+    private boolean compare(final OptionalValueList currents, final OptionalValueList predicates) {
```

Comment:
This private, recursive method has the same name the abstract method that should be implemented. That is confusing. Consider renaming this method `compareList`.